src/odb/src/db/dbBlock.cpp

```diff
@@ -1600,6 +1630,94 @@ void dbBlock::addBTermGroup(const std::vector<dbBTerm*>& bterms, bool order)
   block->_bterm_groups.push_back(std::move(group));
 }
 
+void dbBlock::setBTermTopLayerGrid(int layer_id,
+                                   int x_step,
+                                   int y_step,
+                                   Rect region,
+                                   int pin_width,
+                                   int pin_height,
+                                   int keepout)
+{
+  _dbBlock* block = (_dbBlock*) this;
+  _dbBTermTopLayerGrid& top_layer_grid = block->_bterm_top_layer_grid;
+
+  utl::Logger* logger = block->getImpl()->getLogger();
+
+  const int half_width = pin_width / 2;
+  const int half_height = pin_height / 2;
+  const odb::Rect& die_area = getDieArea();
+  bool region_changed = false;
```

Comment:
I think the db should not be modifying its input and should just error on invalid input.  The user interface can choose to modify the input on the way to the db if it wishes.  

That said I think we are being overly helpful and I don't know that shifting the pin locations is the right thing to do.  Often bump locations are quite fixed by the package design.   @gadfort opinion?

Reply:
I think the out of die area concern should be addressed in ppl, it seems strange to adjust the area on behalf of the user (and strangely couples ppl concerns into odb). ppl can just choose not to allocate slots to pins that would exceed the die area.

Reply:
@gadfort Would you expect PPL to throw a warn message about not using the slots that would exceed the die area? Or should the tool just be silent about it?

Reply:
Silent would be okay. It feels like if you define a region you would expect the pins to be placed inside the region, not straddling the boundary of the region